### ISSUE TITLE
Git mergetool 

### ISSUE THREAD

**Issue by roaladje** [git-for-windows/git#3572]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?
64 bit
```
$ git --version --build-options

** insert your machine's response here **
```
git version 2.34.1.windows.1
cpu: x86_64
built from commit: 2ca94ab318509b3c271e82889938816bad76dfea
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

** insert your machine's response here **
```
Microsoft Windows [Version 10.0.19043.1382]

 - What options did you set as part of the installation? Or did you choose the
   defaults?

``` 
defaults (but notpad++ default editor)

# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

** insert your machine's response here **
```
Editor Option: Notepad++
Custom Editor Path: 
Default Branch Option:  
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

No

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Cmd

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git mergetool
```
 - What did you expect to occur after running these commands?

Merge tool opened and compare files

 - What actually happened instead?
Error:
```
cp: cannot stat './src/dir/file_BACKUP_557.cs': No such file or directory
Normal merge conflict for 'src/dir/file.cs':
  {local}: modified file
  {remote}: modified file
```
Merge tool opened after with "file not found" error.

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Any repo

P.S.
Tested with Meld & BeyondCompare as mergetool and still reproduces. Both tools working as difftool without issues.
Git is set as exclusion in Windows defender (both exploit protection & virus protection)

**Comment by dscho** [comment]
> * What commands did you run to trigger this issue? If you can provide a
>       [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
>       this will help us understand the issue.
> 
> 
> ```
> 1. <merge two branches where there is a merge conflict>
> 2. git mergetool
> ```

Unfortunately, this is several miles away from a verifiable example. If you want anybody to have a prayer to help you, you probably want to read http://stackoverflow.com/help/mcve carefully and put some effort into providing a verifiable example that has a chance of reproducing the issue elsewhere. Otherwise, it is unlikely that anybody can assist you in addressing your troubles.

**Comment by PhilipOakley** [comment]
> * Tested with both Meld & BeyondCompare

Please also include the extra background set ups...

How are these mergetools being setup (e.g. config values)?

Exact copy of the commands being issued to set up the test and the responses (including their paths..)

etc.

**Comment by rajeemm** [comment]
> > * What commands did you run to trigger this issue? If you can provide a
> >   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
> >   this will help us understand the issue.
> > 
> > ```
> > 1. <merge two branches where there is a merge conflict>
> > 2. git mergetool
> > ```
> 
> Unfortunately, this is several miles away from a verifiable example. If you want anybody to have a prayer to help you, you probably want to read http://stackoverflow.com/help/mcve carefully and put some effort into providing a verifiable example that has a chance of reproducing the issue elsewhere. Otherwise, it is unlikely that anybody can assist you in addressing your troubles.








> > * What commands did you run to trigger this issue? If you can provide a
> >   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
> >   this will help us understand the issue.
> > 
> > ```
> > 1. <merge two branches where there is a merge conflict>
> > 2. git mergetool
> > ```
> 
> Unfortunately, this is several miles away from a verifiable example. If you want anybody to have a prayer to help you, you probably want to read http://stackoverflow.com/help/mcve carefully and put some effort into providing a verifiable example that has a chance of reproducing the issue elsewhere. Otherwise, it is unlikely that anybody can assist you in addressing your troubles.

Edited & improved :)


> > * Tested with both Meld & BeyondCompare
> 
> Please also include the extra background set ups...
> 
> How are these mergetools being setup (e.g. config values)?
> 
> Exact copy of the commands being issued to set up the test and the responses (including their paths..)
> 
> etc.

Added

Thanks!

**Comment by PhilipOakley** [comment]
Is your `mergetool` command definition sufficiently complete? i.e. where & how those conflicted & merged files will passed between Git and the Tool?

see https://git-scm.com/docs/git-mergetool#Documentation/git-mergetool.txt-mergetoollttoolgtcmd for the variables and https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration (search for mergetool on the page) for examples.

**Comment by dscho** [comment]
> git init && git checkout -b branch1 && echo foo > 1.txt && git add 1.txt && git commit -m "test conflict" && git checkout -b branch2 && echo bar > 1.txt && git add 1.txt && git commit -m "test conflict" && git checkout branch1 && echo conflict > 1.txt && git add 1.txt && git commit -m "test conflict" && git merge branch2

Unfortunately, I cannot reproduce:

```console
me@work MINGW64 ~/repros/mergetool-cp-problem-3572
$ git init && git checkout -b branch1 && echo foo > 1.txt && git add 1.txt && git commit -m "test conflict" && git checkout -b branch2 && echo bar > 1.txt && git add 1.txt && git commit -m "test conflict" && git checkout branch1 && echo conflict > 1.txt && git add 1.txt && git commit -m "test conflict" && git merge branch2
Initialized empty Git repository in C:/Users/me/repros/mergetool-cp-problem-3572/.git/
Switched to a new branch 'branch1'
warning: LF will be replaced by CRLF in 1.txt.
The file will have its original line endings in your working directory
[branch1 (root-commit) 4707fcc] test conflict
 1 file changed, 1 insertion(+)
 create mode 100644 1.txt
Switched to a new branch 'branch2'
warning: LF will be replaced by CRLF in 1.txt.
The file will have its original line endings in your working directory
[branch2 b57686c] test conflict
 1 file changed, 1 insertion(+), 1 deletion(-)
Switched to branch 'branch1'
warning: LF will be replaced by CRLF in 1.txt.
The file will have its original line endings in your working directory
[branch1 fdc35a2] test conflict
 1 file changed, 1 insertion(+), 1 deletion(-)
Auto-merging 1.txt
CONFLICT (content): Merge conflict in 1.txt
Recorded preimage for '1.txt'
Automatic merge failed; fix conflicts and then commit the result.

me@work MINGW64 ~/repros/mergetool-cp-problem-3572 (branch1|MERGING)
$ git mergetool

This message is displayed because 'merge.tool' is not configured.
See 'git mergetool --tool-help' or 'git help config' for more details.
'git mergetool' will now attempt to use one of the following tools:
opendiff kdiff3 tkdiff xxdiff meld tortoisemerge gvimdiff diffuse diffmerge ecmerge p4merge araxis bc codecompare smerge emerge vimdiff nvimdiff
Merging:
1.txt

Normal merge conflict for '1.txt':
  {local}: modified file
  {remote}: modified file
Hit return to start merge resolution tool (vimdiff):
4 files to edit
```

I see the four files, as expected:

```
  conflict                             |  foo                                   |  bar
  -------------------------------------|  --------------------------------------|  -------------------------------------
  -------------------------------------|  --------------------------------------|  -------------------------------------
  -------------------------------------|  --------------------------------------|  -------------------------------------
  -------------------------------------|  --------------------------------------|  -------------------------------------
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
~                                      |~                                       |~
<14.txt [dos] (11:10 06/12/2021)1,1 All <714.txt [dos] (11:10 06/12/2021)1,1 All <14.txt [dos] (11:10 06/12/2021)1,1 All
  <<<<<<< HEAD
  conflict
  =======
  bar
  >>>>>>> branch2
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
1.txt [dos] (11:10 06/12/2021)                                                                                   1,1 All
"1.txt" [dos] 5L, 55B
```

Now, this is without configuring any mergetool (the `cp` error suggested to me that this was in the shell script part of `git mergetool`, i.e. independent of any configured mergetool). Copying your snippet into `.git/config` fails, though:

```
fatal: bad config line 10 in file .git/config
```

Turns out that I need to double the backslashes:

```ini
[diff]
	tool = bc3
[difftool "bc3"]
	path = C:\\Program Files\\Beyond Compare 4\\BComp.exe
[merge]
	tool = bc3
[mergetool "bc3"]
	path = C:\\Program Files\\Beyond Compare 4\\BComp.exe
```

And when I use that, I get this trace:

```console
me@work MINGW64 ~/repros/mergetool-cp-problem-3572 (branch1|MERGING)
$ git mergetool
Merging:
1.txt

Normal merge conflict for '1.txt':
  {local}: modified file
  {remote}: modified file
1.txt seems unchanged.
Was the merge successful [y/n]?
```

and this window:

<img src=https://user-images.githubusercontent.com/127790/144828468-8b7ae883-c73e-42d6-839c-4a32ae24f44d.png />

Just to make sure, this is the Git version I am using:

```console
me@work MINGW64 ~/repros/mergetool-cp-problem-3572 (branch1|MERGING)
$ git version
git version 2.34.0.windows.1
```

**Comment by rajeemm** [comment]
> Is your `mergetool` command definition sufficiently complete? i.e. where & how those conflicted & merged files will passed between Git and the Tool?
> 
> see https://git-scm.com/docs/git-mergetool#Documentation/git-mergetool.txt-mergetoollttoolgtcmd for the variables and https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration (search for mergetool on the page) for examples.

Yes, I definitely also tried to configure the merge tools with (and with other similar ways) 
```
[mergetool "bc3"]
cmd = \"\<PATH\>\" \"$BASE\" \"$LOCAL\" \"$REMOTE\" \"$MERGED\"
```

Encountering the same issue.

It looks like the issue is before the difftool actually opens.
I assume Git somehow copies the files to be compared to a temporary location.
This is when it fails. 

I able to find a workaround (maybe it will help to understand the issue).
if I'm using Visual Studio to see merge conflicts, and double click the conflict file in VS, the beyond compare (default tool in the PC) opens in mergetool-like state, and works perfectly. I assume this works since it lets VS to create the temporary files and not using Git itself. 


> Now, this is without configuring any mergetool (the `cp` error suggested to me that this was in the shell script part of `git mergetool`, i.e. independent of any configured mergetool). Copying your snippet into `.git/config` fails, though:

Not sure what you mean by this. Isn't Git the one who copies the file to be compared into a temporary location?

**Comment by dscho** [comment]
> Not sure what you mean by this.

I mean: I tried what you described, and cannot reproduce the issue here. Then I went above and beyond by trying a couple things that _might_ be configured on your side but that you did not mention. And I _still_ could not reproduce. In the end, I got too fed up to even try to verify that the report is valid 😁.

**Comment by rajeemm** [comment]
> > Not sure what you mean by this.
> 
> I mean: I tried what you described, and cannot reproduce the issue here. Then I went above and beyond by trying a couple things that _might_ be configured on your side but that you did not mention. And I _still_ could not reproduce. In the end, I got too fed up to even try to verify that the report is valid 😁.

Yes, sorry, about that, I know it will probably not reproduce in all machines, otherwise it would already be solved (my guess is that there is something that clashes with Git, but I have no idea what, and I tried to dig a lot). 
TBH, I tolerate this issue for a long time, since I know it will not consistently reproduce in all machines, so low likeliness of figure it out :(

**Comment by dscho** [comment]
Closing as stale.
